### PR TITLE
Bucket filtering in the functions emulator

### PR DESCRIPTION
**Pull request by colerogers**
Feature change to add bucket filtering to the functions emulator from storage events. Checks if `proto.data.bucket` exists in the multicast handler and appends additional content to the trigger key. This change includes a set of trigger integration tests.

### CODE REVIEW

File: scripts/triggers-end-to-end-tests/tests.ts [hunk 1]
```diff
@@ -295,6 +295,54 @@ describe("storage emulator function triggers", () => {
...
+    this.timeout(EMULATOR_TEST_TIMEOUT);
+
+    const response = await test.writeToSpecificStorageBucket();
+    expect(response.status).to.equal(200);
+    await new Promise((resolve) => setTimeout(resolve, EMULATORS_WRITE_DELAY_MS));
+  });
+
+  it("should have triggered cloud functions", () => {
+    // default bucket values
+    expect(test.storageFinalizedTriggerCount).to.equal(2);
```
Comment (taeold) [inline scripts/triggers-end-to-end-tests/tests.ts:309]: q: it should've not incremented these counts right? Are we just checking here that this values as same as it was before?

(in that case, wow we should update the testing harness to clear the counts between test cases)

**Review by taeold — COMMENTED** [review]
lgtm

File: scripts/triggers-end-to-end-tests/functions/index.js [hunk 2]
```diff
@@ -100,19 +115,33 @@ exports.writeToAuth = functions.https.onRequest(async (req, res) => {
 });
 
 exports.writeToStorage = functions.https.onRequest(async (req, res) => {
-  await admin.storage().bucket().file(STORAGE_FILE_NAME).save("hello world!");
+  await admin.storage().bucket("default-bucket").file(STORAGE_FILE_NAME).save("hello world!");
```
[the remark this answers is not in the document]
Comment (colerogers) [reply scripts/triggers-end-to-end-tests/functions/index.js]: It's not necessary, I added that in to try and debug the tests but ended up being a different issue. I've removed this in the latest push with the working tests

File: scripts/triggers-end-to-end-tests/functions/index.js [hunk 3]
```diff
@@ -175,50 +204,146 @@ exports.authReaction = functions.auth.user().onCreate((user, ctx) => {
   return true;
 });
 
-exports.storageArchiveReaction = functions.storage.object().onArchive((object, context) => {
-  console.log(STORAGE_FUNCTION_ARCHIVED_LOG);
-  console.log("Object", JSON.stringify(object));
-  return true;
-});
+exports.storageArchiveReaction = functions.storage
+  .bucket("default-bucket")
```
[the remark this answers is not in the document]
Comment (colerogers) [reply scripts/triggers-end-to-end-tests/functions/index.js]: Removed, see above comment

File: scripts/triggers-end-to-end-tests/tests.ts (hunk 1 above)
Comment (colerogers) [reply scripts/triggers-end-to-end-tests/tests.ts]: Yeah I was just checking the values from the previous run, I added in a `resetCounts` function in the framework to clean this up a bit

File: scripts/triggers-end-to-end-tests/tests.ts [hunk 4]
```diff
@@ -261,39 +261,114 @@ describe("storage emulator function triggers", () => {
...
+    expect(test.storageDeletedTriggerCount).to.equal(0);
+    expect(test.storageV2DeletedTriggerCount).to.equal(0);
+    // specific bucket
+    expect(test.storageBucketFinalizedTriggerCount).to.equal(1);
+    expect(test.storageBucketMetadataTriggerCount).to.equal(1);
+    expect(test.storageBucketV2FinalizedTriggerCount).to.equal(1);
+    expect(test.storageBucketV2MetadataTriggerCount).to.equal(1);
+    expect(test.storageBucketDeletedTriggerCount).to.equal(0);
+    expect(test.storageBucketV2DeletedTriggerCount).to.equal(0);
+    test.resetCounts();
```
Comment (taeold) [inline scripts/triggers-end-to-end-tests/tests.ts:315]: aside* I'd like to see this function be called in `afterEach` hook, but I see that this isn't possible due to the way the test cases are structured today :/.

**Review by taeold — APPROVED** [review]
lgtm, but would love to have @abeisgoat review.

**Comment by abeisgoat** [comment]
Love the comprehensive tests, looks good to me.

**Review by abeisgoat — APPROVED** [review]

**Pull request merged**